File: sdk/typescript/tests/parser.test.ts
import { describe, it, expect, beforeAll } from 'vitest';
import { parse, parseHL7, parseCSV, parseCSVWithSchema, FiFhirError, isFiFhirAvailable } from '../src';

describe('Parser', () => {
  let fiFhirAvailable: boolean;

  beforeAll(async () => {
    fiFhirAvailable = await isFiFhirAvailable();
  });

  describe('parseHL7', () => {
    it.skipIf(!fiFhirAvailable)('parses ADT^A01 message', async () => {
      const message = `MSH|^~\\&|EPIC|HOSPITAL|DEST|DEST|20240115120000||ADT^A01|MSG001|P|2.5
EVN|A01|20240115120000
PID|1||123456^^^HOSP^MR||DOE^JOHN^W||19800315|M|||123 MAIN ST^^ANYTOWN^VA^24101||555-123-4567
PV1|1|I|ICU^101^A^HOSP||||12345^SMITH^ROBERT|||MED||||||||V123456^^^HOSP^VN|||||||||||||||||||||||||20240115110000`;

      const event = await parseHL7(message, { source: 'test_system' });

      expect(event.type).toBe('patient_admit');
      expect(event.source).toBe('test_system');
      expect(event.source_format).toBe('hl7v2');

      if (event.type === 'patient_admit') {
        expect(event.patient.mrn).toBe('123456');
        expect(event.patient.family_name).toBe('DOE');
        expect(event.patient.given_name).toBe('JOHN');
        expect(event.patient.gender).toBe('M');
      }
    });

    it.skipIf(!fiFhirAvailable)('parses ORU^R01 lab result', async () => {
      const message = `MSH|^~\\&|LAB|HOSPITAL|DEST|DEST|20240115140000||ORU^R01|MSG002|P|2.5
PID|1||123456^^^HOSP^MR||DOE^JOHN
OBR|1|ORD001|ACC001|CBC^Complete Blood Count|||20240115130000
OBX|1|NM|WBC^White Blood Cell Count||12.5|10*3/uL|4.5-11.0|H|||F`;

      const event = await parseHL7(message);

      expect(event.type).toBe('lab_result');

      if (event.type === 'lab_result') {
        expect(event.patient.mrn).toBe('123456');
        expect(event.test.local_code).toBe('WBC');
        expect(event.result.value).toBe('12.5');
        expect(event.result.interpretation).toBe('H');
      }
    });
  });

  describe('parseCSV', () => {
    it.skipIf(!fiFhirAvailable)('parses patient CSV', async () => {
      const csv = `mrn,first_name,last_name,dob,gender
123456,John,Doe,1980-03-15,M
789012,Jane,Smith,1992-07-22,F`;

      const events = await parseCSV(csv, { eventType: 'patient' });

      expect(events).toHaveLength(2);
      expect(events[0].type).toBe('patient_update');

      if (events[0].type === 'patient_update') {
        expect(events[0].patient.mrn).toBe('123456');
        expect(events[0].patient.given_name).toBe('John');
        expect(events[0].patient.family_name).toBe('Doe');
      }
    });

    it.skipIf(!fiFhirAvailable)('parses lab result CSV', async () => {
      const csv = `mrn,test_code,test_name,result,unit,interpretation
123456,GLU,Glucose,95,mg/dL,N
123456,HGB,Hemoglobin,14.2,g/dL,N`;

      const events = await parseCSV(csv, { eventType: 'lab' });

      expect(events).toHaveLength(2);
      expect(events[0].type).toBe('lab_result');

      if (events[0].type === 'lab_result') {
        expect(events[0].test.local_code).toBe('GLU');
        expect(events[0].result.value).toBe('95');
      }
    });

    it.skipIf(!fiFhirAvailable)('handles custom delimiter', async () => {
      const tsv = `mrn\tfirst_name\tlast_name
123456\tJohn\tDoe`;

      const events = await parseCSV(tsv, {
        eventType: 'patient',
        delimiter: 'tab'
      });

      expect(events).toHaveLength(1);
      if (events[0].type === 'patient_update') {
        expect(events[0].patient.mrn).toBe('123456');
      }
    });
  });

  describe('parseCSVWithSchema', () => {
    it.skipIf(!fiFhirAvailable)('infers schema from CSV', async () => {
      const csv = `mrn,first_name,last_name,dob,gender,ssn,phone,email
123456,John,Doe,1980-03-15,M,123-45-6789,555-123-4567,john@example.com
789012,Jane,Smith,1992-07-22,F,987-65-4321,555-987-6543,jane@example.com`;

      const result = await parseCSVWithSchema(csv);

      expect(result.schema).toBeDefined();
      expect(result.schema?.columns).toHaveLength(8);

      const mrnCol = result.schema?.columns.find(c => c.name === 'mrn');
      expect(mrnCol?.inferred_type).toBe('mrn');
      expect(mrnCol?.semantic_hint).toBe('patient_mrn');

      const ssnCol = result.schema?.columns.find(c => c.name === 'ssn');
      expect(ssnCol?.inferred_type).toBe('ssn');

      const emailCol = result.schema?.columns.find(c => c.name === 'email');
      expect(emailCol?.inferred_type).toBe('email');
    });
  });

  describe('error handling', () => {
    it.skipIf(!fiFhirAvailable)('throws FiFhirError for invalid input', async () => {
      await expect(parseCSV('', { eventType: 'patient' }))
        .rejects.toThrow(FiFhirError);
    });

    it('throws FiFhirError when binary not found', async () => {
      // This test will only work if fi-fhir is NOT installed
      // Skip in CI where it's available
      if (fiFhirAvailable) {
        return;
      }

      await expect(parseHL7('invalid'))
        .rejects.toThrow(/fi-fhir binary not found/);
    });
  });
});

File: internal/parser/csv/parser_test.go
package csv

import (
	"testing"

	"github.com/cblevins/fi-fhir/pkg/events"
)

func TestParsePatientCSV(t *testing.T) {
	parser := NewParser("test_source", ParserConfig{
		HasHeader: true,
		EventType: events.EventPatientUpdate,
	})

	csv := `mrn,first_name,last_name,dob,gender,address,city,state,zip,phone
123456,John,Doe,1980-03-15,M,123 Main St,Anytown,VA,24101,555-123-4567
789012,Jane,Smith,1992-07-22,F,456 Oak Ave,Springfield,IL,62701,555-987-6543`

	result, err := parser.ParseString(csv)
	if err != nil {
		t.Fatalf("Parse failed: %v", err)
	}

	if len(result.Events) != 2 {
		t.Fatalf("Expected 2 events, got %d", len(result.Events))
	}

	// Check first patient
	event1 := result.Events[0].(*events.PatientAdmitEvent)
	if event1.Patient.MRN != "123456" {
		t.Errorf("Expected MRN '123456', got '%s'", event1.Patient.MRN)
	}
	if event1.Patient.GivenName != "John" {
		t.Errorf("Expected given name 'John', got '%s'", event1.Patient.GivenName)
	}
	if event1.Patient.FamilyName != "Doe" {
		t.Errorf("Expected family name 'Doe', got '%s'", event1.Patient.FamilyName)
	}
	if event1.Patient.Gender != "M" {
		t.Errorf("Expected gender 'M', got '%s'", event1.Patient.Gender)
	}
	if event1.Patient.Address.City != "Anytown" {
		t.Errorf("Expected city 'Anytown', got '%s'", event1.Patient.Address.City)
	}

	// Check second patient
	event2 := result.Events[1].(*events.PatientAdmitEvent)
	if event2.Patient.MRN != "789012" {
		t.Errorf("Expected MRN '789012', got '%s'", event2.Patient.MRN)
	}
}

func TestParseLabResultCSV(t *testing.T) {
	parser := NewParser("lab_system", ParserConfig{
		HasHeader: true,
		EventType: events.EventLabResult,
	})

	csv := `mrn,patient_last_name,patient_first_name,test_code,test_name,result,unit,reference_range,interpretation,collection_date
123456,Doe,John,GLU,Glucose,95,mg/dL,70-100,N,2024-01-15
123456,Doe,John,HGB,Hemoglobin,14.2,g/dL,12.0-17.5,N,2024-01-15
789012,Smith,Jane,GLU,Glucose,142,mg/dL,70-100,H,2024-01-16`

	result, err := parser.ParseString(csv)
	if err != nil {
		t.Fatalf("Parse failed: %v", err)
	}

	if len(result.Events) != 3 {
		t.Fatalf("Expected 3 events, got %d", len(result.Events))
	}

	// Check first lab result
	event1 := result.Events[0].(*events.LabResultEvent)
	if event1.Patient.MRN != "123456" {
		t.Errorf("Expected MRN '123456', got '%s'", event1.Patient.MRN)
	}
	if event1.Test.LocalCode != "GLU" {
		t.Errorf("Expected test code 'GLU', got '%s'", event1.Test.LocalCode)
	}
	if event1.Result.Value != "95" {
		t.Errorf("Expected result '95', got '%s'", event1.Result.Value)
	}
	if event1.Result.Unit != "mg/dL" {
		t.Errorf("Expected unit 'mg/dL', got '%s'", event1.Result.Unit)
	}
	if event1.Result.Interpretation != "N" {
		t.Errorf("Expected interpretation 'N', got '%s'", event1.Result.Interpretation)
	}

	// Check high glucose
	event3 := result.Events[2].(*events.LabResultEvent)
	if event3.Result.Interpretation != "H" {
		t.Errorf("Expected interpretation 'H', got '%s'", event3.Result.Interpretation)
	}
}

func TestSchemaInference(t *testing.T) {
	parser := NewParser("test", ParserConfig{
		HasHeader:   true,
		InferSchema: true,
	})

	csv := `mrn,first_name,last_name,dob,gender,ssn,phone,email
123456,John,Doe,1980-03-15,M,123-45-6789,555-123-4567,john@example.com
789012,Jane,Smith,1992-07-22,F,987-65-4321,555-987-6543,jane@example.com
345678,Bob,Wilson,1975-11-30,M,456-78-9012,555-456-7890,bob@example.com`

	result, err := parser.ParseString(csv)
	if err != nil {
		t.Fatalf("Parse failed: %v", err)
	}

	if result.Schema == nil {
		t.Fatal("Expected schema to be inferred")
	}

	// Check inferred types
	expectedTypes := map[string]ColumnType{
		"mrn":        TypeMRN,
		"first_name": TypeString,
		"last_name":  TypeString,
		"dob":        TypeDate,
		"gender":     TypeGender,
		"ssn":        TypeSSN,
		"phone":      TypePhone,
		"email":      TypeEmail,
	}

	for _, col := range result.Schema.Columns {
		if expected, ok := expectedTypes[col.Name]; ok {
			if col.InferredType != expected {
				t.Errorf("Column %s: expected type %s, got %s", col.Name, expected, col.InferredType)
			}
		}
	}
}

func TestGenderNormalization(t *testing.T) {
	parser := NewParser("test", ParserConfig{})

	tests := []struct {
		input    string
		expected string
	}{
		{"M", "M"},
		{"F", "F"},
		{"male", "M"},
		{"FEMALE", "F"},
		{"1", "M"},
		{"2", "F"},
		{"Other", "O"},
		{"unknown", "U"},
		{"", "U"},
	}

	for _, tt := range tests {
		result := parser.normalizeGender(tt.input)
		if result != tt.expected {
			t.Errorf("normalizeGender(%q) = %q, want %q", tt.input, result, tt.expected)
		}
	}
}

func TestDateParsing(t *testing.T) {
	parser := NewParser("test", ParserConfig{})

	tests := []struct {
		input string
		valid bool
	}{
		{"2024-01-15", true},
		{"01/15/2024", true},
		{"01-15-2024", true},
		{"2024/01/15", true},
		{"20240115", true},
		{"not a date", false},
	}

	for _, tt := range tests {
		_, err := parser.parseDate(tt.input)
		if (err == nil) != tt.valid {
			t.Errorf("parseDate(%q): valid=%v, want valid=%v", tt.input, err == nil, tt.valid)
		}
	}
}

func TestCustomDelimiter(t *testing.T) {
	parser := NewParser("test", ParserConfig{
		HasHeader: true,
		Delimiter: '\t',
		EventType: events.EventPatientUpdate,
	})

	tsv := "mrn\tfirst_name\tlast_name\n123456\tJohn\tDoe"

	result, err := parser.ParseString(tsv)
	if err != nil {
		t.Fatalf("Parse failed: %v", err)
	}

	if len(result.Events) != 1 {
		t.Fatalf("Expected 1 event, got %d", len(result.Events))
	}

	event := result.Events[0].(*events.PatientAdmitEvent)
	if event.Patient.MRN != "123456" {
		t.Errorf("Expected MRN '123456', got '%s'", event.Patient.MRN)
	}
}

func TestColumnMapping(t *testing.T) {
	parser := NewParser("test", ParserConfig{
		HasHeader: true,
		EventType: events.EventPatientUpdate,
		ColumnMapping: map[string]string{
			"pat_id": "mrn",
			"fname":  "first_name",
			"lname":  "last_name",
		},
	})

	csv := `pat_id,fname,lname,dob
123456,John,Doe,1980-03-15`

	result, err := parser.ParseString(csv)
	if err != nil {
		t.Fatalf("Parse failed: %v", err)
	}

	event := result.Events[0].(*events.PatientAdmitEvent)
	if event.Patient.MRN != "123456" {
		t.Errorf("Expected MRN '123456', got '%s'", event.Patient.MRN)
	}
	if event.Patient.GivenName != "John" {
		t.Errorf("Expected given name 'John', got '%s'", event.Patient.GivenName)
	}
}

func TestEmptyCSV(t *testing.T) {
	parser := NewParser("test", ParserConfig{})

	_, err := parser.ParseString("")
	if err == nil {
		t.Error("Expected error for empty CSV")
	}
}

func TestNoHeader(t *testing.T) {
	parser := NewParser("test", ParserConfig{
		HasHeader: false,
		ColumnMapping: map[string]string{
			"col_0": "mrn",
			"col_1": "first_name",
			"col_2": "last_name",
		},
		EventType: events.EventPatientUpdate,
	})

	csv := `123456,John,Doe
789012,Jane,Smith`

	result, err := parser.ParseString(csv)
	if err != nil {
		t.Fatalf("Parse failed: %v", err)
	}

	if len(result.Events) != 2 {
		t.Fatalf("Expected 2 events, got %d", len(result.Events))
	}

	event := result.Events[0].(*events.PatientAdmitEvent)
	if event.Patient.MRN != "123456" {
		t.Errorf("Expected MRN '123456', got '%s'", event.Patient.MRN)
	}
}

func TestSemanticHints(t *testing.T) {
	parser := NewParser("test", ParserConfig{})

	tests := []struct {
		name     string
		colType  ColumnType
		expected string
	}{
		{"mrn", TypeMRN, "patient_mrn"},
		{"patient_id", TypeString, "patient_mrn"},
		{"first_name", TypeString, "patient_given_name"},
		{"dob", TypeDate, "patient_dob"},
		{"loinc_code", TypeString, "lab_loinc_code"},
		{"icd10", TypeCode, "diagnosis_icd_code"},
		{"random_column", TypeString, ""},
	}

	for _, tt := range tests {
		hint := parser.inferSemanticHint(tt.name, tt.colType)
		if hint != tt.expected {
			t.Errorf("inferSemanticHint(%q, %s) = %q, want %q", tt.name, tt.colType, hint, tt.expected)
		}
	}
}

func TestInferColumnType(t *testing.T) {
	parser := NewParser("test", ParserConfig{})

	tests := []struct {
		name     string
		samples  []string
		expected ColumnType
	}{
		{"ssn_col", []string{"123-45-6789", "987-65-4321", "456-78-9012"}, TypeSSN},
		{"email_col", []string{"a@b.com", "x@y.org", "test@example.com"}, TypeEmail},
		{"gender_col", []string{"M", "F", "M", "F"}, TypeGender},
		{"date_col", []string{"2024-01-15", "2024-02-20", "2024-03-25"}, TypeDate},
		{"int_col", []string{"123", "456", "789"}, TypeInteger},
		{"float_col", []string{"1.5", "2.7", "3.14"}, TypeFloat},
		{"code_col", []string{"A123", "B456", "C789.1"}, TypeCode},
		{"mixed", []string{"abc", "123", "def"}, TypeString},
	}

	for _, tt := range tests {
		result := parser.inferColumnType(tt.name, tt.samples)
		if result != tt.expected {
			t.Errorf("inferColumnType(%q, %v) = %s, want %s", tt.name, tt.samples, result, tt.expected)
		}
	}
}
